bzzeth/bzzeth.go

```diff
@@ -326,6 +372,256 @@ func newChunk(data []byte) chunk.Chunk {
+func (b *BzzEth) getBlockHeaderBzz(ctx context.Context, hash []byte) ([]byte, error) {
+	req := &storage.Request{
+		Addr:   hash,
+		Origin: b.netStore.LocalID,
+	}
+	chnk, err := b.netStore.Get(ctx, chunk.ModeGetRequest, req)
+	if err != nil {
+		return nil, err
+	}
+	return chnk.Data(), nil
+}
+
+// requestAll requests each hash and channel
+func (b *BzzEth) requestAll(ctx context.Context, deliveries chan []byte, hashes [][]byte, rcvdPeer *Peer) {
+	ctx, cancel := context.WithTimeout(ctx, timeouts.FetcherGlobalTimeout)
+	defer cancel()
+
+	// missingHeaders collects hashes of headers not found within swarm
+	// ie., the hashes to request from the eth full nodes
+	missingHeaders := make(chan []byte)
+	defer close(missingHeaders)
+	var wg sync.WaitGroup
+
+BZZ:
+	for _, h := range hashes {
+		hdr := make([]byte, len(h))
+		copy(hdr, h)
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+			header, err := b.getBlockHeaderBzz(ctx, hdr)
+			if err != nil {
+				log.Debug("bzzeth.requestAll: netstore.Get can not retrieve chunk", "ref", hex.EncodeToString(h), "err", err)
+				select {
+				case missingHeaders <- hdr: // fallback: request header from eth peers
+				case <-ctx.Done():
+				}
+				return
+			}
+			// deliver the headers received from Swarm
+			select {
+			case deliveries <- header:
+			case <-ctx.Done():
+			}
+		}()
+
+		select {
+		case <-ctx.Done():
+			break BZZ
+		default:
+		}
+	}
+
+	// fall back to retrieval from eth clients
+	// collect missing block header hashes
+	// terminates after missingHeaders is read and closed or context is done
+	wg.Add(1)
+	go b.getBlockHeadersEth(ctx, missingHeaders, deliveries, &wg, rcvdPeer)
+
+	// wait till all hashes are requested from swarm OR from another eth node,
+	wg.Wait()
+
+}
+
+// getBlockHeadersEth manages fetching headers from ethereum bzzeth nodes
+// This is part of the response to GetBlockHeaders requests by bzzeth light/syncing nodes
+// As a fallback after header retrieval from local storage and swarm network are unsuccessful
+// When called, it
+// - reads requested header hashes from a channel (headerC) and
+// - creates batch requests and sends them to an adequate bzzeth peer
+// - channels the responses into a delivery channel (deliveries)
+func (b *BzzEth) getBlockHeadersEth(ctx context.Context, headersC, giveBackC chan []byte, pwg *sync.WaitGroup, rcvdPeer *Peer) {
+	log.Debug("getting missing headers from another ETH node", "count", len(giveBackC))
+	defer pwg.Done() // unblock the parent so that i can continue
+
+	// read header requests into batches
+	readNext := make(chan chan [][]byte)
+	batches := make(chan [][]byte)
+	go readToBatches(headersC, readNext)
+	readNext <- batches
+
+	// send GetBlockHeader requests to adequate bzzeth peers
+	// this loop terminates when batches channel is closed as a result of input headersC being closed
+	requiredCount := 0
+	var wg sync.WaitGroup
+	deliveryC := make(chan []byte)
```

Comment:
why this other channel?

Reply:
explained above.

Reply:
still dont get it, you just need one for requests the other for responses that you then sent back to the requesting peer after validation and storage.